File: day30/alien_dictionary.py
# defaultdict lets us automatically create a set for each key without needing to check if the key exists.
# deque is a double-ended queue optimized for popping from the left, perfect for BFS.
from collections import defaultdict, deque

def alienOrder(words):
    # Step 1: Build graph and in-degree map
    # Each character will have an entry in in_degree initialized to 0
    graph = defaultdict(set)
    in_degree = {char: 0 for word in words for char in word}

    print("Initial in-degree map:", in_degree)

    # Compare adjacent words to build the graph
    for i in range(len(words) - 1):
        first, second = words[i], words[i+1]
        for a, b in zip(first, second):
            if a != b:
                # Only add edge if it's not already present
                if b not in graph[a]:
                    graph[a].add(b)
                    in_degree[b] += 1
                break
        else:
            # Handle invalid prefix case (e.g., "abc" before "ab")
            if len(second) < len(first):
                return ""

    # Debug print of the graph
    print("Graph (edges):")
    for k, v in graph.items():
        print(f"  {k} → {v}")
    print("Updated in-degree map after edges:", in_degree)

    # Step 2: Topological sort using BFS
    # Start with all nodes that have 0 in-degree
    queue = deque([c for c in in_degree if in_degree[c] == 0])
    result = []

    print("Starting BFS with queue:", list(queue))

    while queue:
        c = queue.popleft()
        result.append(c)
        print(f"Visiting: {c}")
        for neighbor in graph[c]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    print("Final result list:", result)

    # If not all letters are in result, a cycle was detected
    if len(result) < len(in_degree):
        return ""

    return "".join(result)

# Example usage
words = ["wrt", "wrf", "er", "ett", "rftt"]
print("Input words:", words)
print("Alien dictionary order:", alienOrder(words))

File: day30/alien_dictionary.rs
use std::collections::{HashMap, HashSet, VecDeque};

fn alien_order(words: Vec<&str>) -> Option<String> {
    let mut graph: HashMap<char, HashSet<char>> = HashMap::new();
    let mut in_degree: HashMap<char, usize> = HashMap::new();

    println!("Input words: {:?}", words);

    // Initialize in-degree
    for word in &words {
        for c in word.chars() {
            in_degree.entry(c).or_insert(0);
        }
    }

    println!("Initial in-degree map: {:?}", in_degree);

    // Build graph
    for pair in words.windows(2) {
        let (first, second) = (pair[0], pair[1]);
        let mut found = false;
        for (c1, c2) in first.chars().zip(second.chars()) {
            if c1 != c2 {
                let entry = graph.entry(c1).or_default();
                if entry.insert(c2) {
                    *in_degree.entry(c2).or_insert(0) += 1;
                }
                found = true;
                break;
            }
        }
        if !found && first.len() > second.len() {
            return None;
        }
    }

    println!("Graph (edges):");
    for (k, v) in &graph {
        let edges: Vec<char> = v.iter().cloned().collect();
        println!("  {} → {:?}", k, edges);
    }

    println!("Updated in-degree map: {:?}", in_degree);

    let mut queue = VecDeque::new();
    for (&c, &deg) in &in_degree {
        if deg == 0 {
            queue.push_back(c);
        }
    }

    println!("Starting BFS with queue: {:?}", queue);

    let mut result = String::new();
    while let Some(c) = queue.pop_front() {
        println!("Visiting: {}", c);
        result.push(c);
        if let Some(neighbors) = graph.get(&c) {
            for &n in neighbors {
                if let Some(d) = in_degree.get_mut(&n) {
                    *d -= 1;
                    if *d == 0 {
                        queue.push_back(n);
                    }
                }
            }
        }
    }

    if result.len() == in_degree.len() {
        println!("Alien dictionary order: {}", result);
        Some(result)
    } else {
        println!("Invalid input — no valid order exists.");
        None
    }
}

fn main() {
    alien_order(vec!["wrt", "wrf", "er", "ett", "rftt"]);
}
